File: packages/code-runtime/code-runtime-worker/src/protocol.ts
/**
 * Versionless, structured-clone wire protocol between co-shipped host and worker code. The host
 * treats inbound traffic as hostile because model code can forge `parentPort` messages; the
 * worker trusts host replies.
 * @module @deepseek-ai/dsh-code-runtime-worker/src/protocol
 */

/** What the host hands the worker at spawn, via `workerData`. */
export interface WorkerBootData {
  /** The type-stripped (plain JS) program body. */
  code: string
  /** Binding namespaces to materialize: the global name plus the function names (functions themselves stay host-side). */
  namespaces: { global: string; names: string[] }[]
  /** Shared byte budget for captured log text; exceeding it drops further entries after one in-band marker. */
  maxLogBytes: number
  /** Byte cap for the rendered completion value (see the value-preparation contract in bootstrap.ts). */
  maxValueBytes: number
}

/** Worker → host: one bridged binding call. */
interface CallMessage {
  type: 'call'
  /** Worker-issued correlation id; the host answers each id at most once and ignores duplicates. */
  id: number
  /** The namespace global the call targets. */
  global: string
  /** The function name within the namespace. */
  name: string
  /** The single argument, structured-clone-plain. */
  args: unknown
}

/** Worker → host: captured text, streamed eagerly so output survives a mid-run termination (timeout, abort, OOM). */
interface LogMessage {
  type: 'log'
  text: string
}

/**
 * Worker → host: the program settled. `error` carries a program exception
 * (the only failure the bootstrap itself can report — budgets, aborts, and
 * substrate death are observed host-side). `value` is present only on a
 * clean completion that produced one (already size-capped and
 * clone-safe per the bootstrap's value preparation). Logs are NOT carried
 * here — they streamed eagerly as {@link LogMessage}s.
 */
export interface DoneMessage {
  type: 'done'
  value?: unknown
  error?: { message: string }
}

/** Every message the worker sends. */
export type WorkerToHost = CallMessage | LogMessage | DoneMessage

/** Host → worker: the answer to one {@link CallMessage}. */
export type ReplyMessage =
  | { type: 'reply'; id: number; ok: true; value: unknown }
  | { type: 'reply'; id: number; ok: false; message: string }

/**
 * The in-band marker entry text announcing that log capture stopped at the
 * byte budget. Shared wire vocabulary: the worker's LogBuffer emits it when
 * ITS budget exhausts, and the host emits the identical text when its own
 * ledger drops an entry first (forged port traffic, stray pipe bytes) — so
 * a truncated run reads the same however the cap was hit.
 * @param maxBytes - the configured `maxLogBytes` the marker names.
 * @returns the marker line.
 */
export function logTruncationMarker(maxBytes: number): string {
  return `[dsh-code-runtime-worker] log capture truncated at ${maxBytes} bytes`
}

File: packages/code-runtime/code-runtime-python/py/protocol.py
"""Wire protocol vocabulary for the Python side of dsh-code-runtime-python.

Mirrors ``src/protocol.ts``. Frames travel on fd 3 as JSON-lines (one JSON
object per line). The host validates every inbound frame; this side trusts
host replies.
"""

from __future__ import annotations

from typing import Any, Literal, TypedDict, Union

# The protocol fd from the child's perspective. Node passes
# ``stdio: [pipe, pipe, pipe, pipe]`` so the fourth entry (fd 3) is the
# framed-JSON channel; stdout/stderr stay clear for the program's own output.
PROTOCOL_FD = 3


class BootMessage(TypedDict):
    """Host → child, first frame on fd 3. Carries every cap and the namespaces."""

    type: Literal["boot"]
    cpuSeconds: int
    addressSpaceBytes: int
    maxLogBytes: int
    maxValueBytes: int
    namespaces: list["Namespace"]


class ErrorClass(TypedDict):
    """A namespace's program-visible exception class: rejected calls raise its
    instances carrying the failed member name on ``memberNameProperty``."""

    name: str
    memberNameProperty: str


class Namespace(TypedDict, total=False):
    """One binding namespace declaration: the global name, its function names,
    and an optional program-visible ``errorClass`` for rejected calls."""

    global_: str  # required; renamed on the wire: JSON field is ``global`` (Python keyword collision)
    names: list[str]  # required
    errorClass: ErrorClass  # optional — mirrors the TS `errorClass?`


class RunMessage(TypedDict):
    """Host → child, sent after ``boot-ack``. Carries only the program body."""

    type: Literal["run"]
    program: str


class BootAckMessage(TypedDict):
    """Child → host: resource limits applied, ready for the run message."""

    type: Literal["boot-ack"]


class CallMessage(TypedDict):
    """Child → host: one bridged binding call from the model program."""

    type: Literal["call"]
    id: int
    global_: str  # wire field is ``global``
    name: str
    args: Any


class LogMessage(TypedDict, total=False):
    """Child → host: one captured text chunk, streamed eagerly.

    ``truncated`` is set only on the frame that IS the child ledger's truncation
    marker (not program output), so the host stops capturing at the same point
    the child did — mirrors the TS `truncated?`.
    """

    type: Literal["log"]  # required
    text: str  # required
    truncated: bool  # optional


class DoneErrorField(TypedDict):
    """Child → host: the failure carried on a ``done`` frame. ``kind`` is one of
    the three the host validates; ``message`` is the traceback or diagnostic."""

    kind: Literal["exception", "invalid-output", "output-limit"]
    message: str


class DoneMessage(TypedDict, total=False):
    """Child → host: the program settled. ``value`` and ``error`` are optional per the TS mirror."""

    type: Literal["done"]  # required — TypedDict(total=False) allows this via a required subclass in Py 3.11+; MVP keeps it flat
    value: Any
    error: DoneErrorField


ChildToHost = Union[BootAckMessage, CallMessage, LogMessage, DoneMessage]


class ReplyOk(TypedDict):
    type: Literal["reply"]
    id: int
    ok: Literal[True]
    value: Any


class ReplyErr(TypedDict):
    type: Literal["reply"]
    id: int
    ok: Literal[False]
    message: str


ReplyMessage = Union[ReplyOk, ReplyErr]
HostToChild = ReplyMessage


def log_truncation_marker(max_bytes: int) -> str:
    """Return the in-band marker for a log ledger that exhausted its budget.

    Byte-identical text on both sides of the wire so a truncated run reads the
    same however the cap was hit.
    """

    return f"[dsh-code-runtime-python] log capture truncated at {max_bytes} bytes"
